driven/generic/adapter.py

```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
+    If metabolite does not exist in the BIGG database, use Metanetx id.
+    If compartment is given, metabolites ids will have it as postfix.
+
+    Example:
+    Input: C00002 + C00033 <=> C00013 + C05993, compartment='_c'
+    Output: atp_c + ac_c <=> ppi_c + MNXM4377_c
+
+    :param equation: string
+    :param compartment: f.e. "_c"
+    :return:
+    """
+    array = equation.split()
+    result = []
+    for i, el in enumerate(array):
+        if not re.match("^[A-Za-z][A-Za-z0-9]*$", el):
+            result.append(el)
+        else:
+            try:
+                el = metanetx.all2mnx['kegg:' + el]
+                el = metanetx.mnx2bigg[el].replace('bigg:', '')
+            except KeyError:
+                pass
+            if compartment:
+                el += compartment
+            result.append(el)
+    return ' '.join(result)
+
+
+def full_genotype(genotype_changes):
+    """Construct gnomic Genotype object from the list of strings with changes
+
+    :param genotype_changes: list of changes, f.e. ['-tyrA::kanMX+', 'kanMX-']
+    :return:
+    """
+
+    def chain(definitions, **kwargs):
+        if not definitions:
+            return gnomic.Genotype([])
+        genotype = gnomic.Genotype.parse(definitions[0], **kwargs)
+        for definition in definitions[1:]:
+            genotype = gnomic.Genotype.parse(definition, parent=genotype, **kwargs)
+        return genotype
+
+    return chain(genotype_changes)
+
+
+class GenotypeChangeModel(ModelModificationMixin):
+    """
+    Applies genotype change on cameo model
+    """
+
+    def __init__(self, model, genotype_changes, genes_to_reactions):
+        """Initialize change model
+
+        :param model: cameo model
+        :param genotype_changes: gnomic.Genotype object
+        :param genes_to_reactions: dictionary like {<gene name>: {<reaction id>: <reactions equation>, ...}, ...}
+        """
+        self.compartment = '_c'
+        self.model = model
+        self.genes_to_reactions = genes_to_reactions
+        self.changes = {
+            'added': {'reactions': set(), 'metabolites': set()},  # reaction contain information about genes
+            'removed': {'genes': set()},
+        }
+        self.apply_changes(genotype_changes)
+
+    def apply_changes(self, genotype_changes):
+        """Apply genotype changes on initial model
+
+        :param genotype_changes: gnomic.Genotype
+        :return:
+        """
+        for change in genotype_changes.changes():
+            if isinstance(change, gnomic.Mutation):
+                self.apply_mutation(change)
+            if isinstance(change, gnomic.Plasmid):
+                self.add_plasmid(change)
+
+    def apply_mutation(self, mutation):
+        """Apply mutations on initial model
+
+        :param mutation: gnomic.Mutation
+        :return:
+        """
+        if mutation.old:
+            for feature in mutation.old.features():
+                self.knockout_gene(feature)
```

Comment:
Is the assumption here that mutations will always knockout the gene? Shouldn't it say something about that in the documentation? What about mutations that change the activity, like constraining it to half?